src/Orleans/CodeGeneration/TypeUtils.cs

```diff
@@ -324,22 +334,47 @@ public static bool IsSystemTargetClass(Type type)
                 systemTargetBaseInterfaceType = ToReflectionOnlyType(systemTargetBaseInterfaceType);
             }
 
-            if (!systemTargetInterfaceType.IsAssignableFrom(type) || 
-                !systemTargetBaseInterfaceType.IsAssignableFrom(type) || 
-                !systemTargetType.IsAssignableFrom(type)) return false;
+            if (!systemTargetInterfaceType.IsAssignableFrom(type) ||
+                !systemTargetBaseInterfaceType.IsAssignableFrom(type) ||
+                !systemTargetType.IsAssignableFrom(type))
+            {
+                if (complain)
+                {
+                    violations = new string[] { String.Format("Class {0} extends IGrain but is not a concrete Grain class or a System Target. Hint: Extend from base Grain or Grain<T>.", type.FullName) };
+                }
+                return false;
+            }
 
             // exclude generated classes.
-            return !IsGeneratedType(type);
+            if (IsGeneratedType(type))
+            {
+                if (complain)
+                {
+                    violations = new string[] { String.Format("Ignoring {0}, as it is an auto-generated class.", type.FullName) };
+                }
+                return false;
+            }
+            return true;
         }
 
         public static bool IsConcreteGrainClass(Type type, out IEnumerable<string> complaints, bool complain)
```

Comment:
I guess we are already violating this logic here and in other places. Then it's fine to take your change, and refactor these `IsX()` methods as a separate change.